packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -0,0 +1,461 @@
+        for (uint i = 0; i < tokens.length; i++) {
+            if (ammBalances[i] > initialAmmBalances[i]) {
+                uint amount = ammBalances[i] - initialAmmBalances[i];
+                // Check that the deposit in the block matches the expected deposit
+                DepositTransaction.Deposit memory _deposit = _block.readDeposit(txIdx++);
+                require(_deposit.owner == address(this), "INVALID_TX_DATA");
+                require(_deposit.accountID == accountID, "INVALID_TX_DATA");
+                require(_deposit.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+                require(_deposit.amount == amount, "INVALID_TX_DATA");
+                // Now do this deposit
+                uint ethValue = 0;
+                if (tokens[i].addr == address(0)) {
+                    ethValue = _deposit.amount;
+                } else {
+                    // Approve the deposit transfer
+                    ERC20(tokens[i].addr).approve(address(exchange), _deposit.amount);
+                }
+                exchange.deposit{value: ethValue}(
+                    _deposit.owner,
+                    _deposit.owner,
+                    tokens[i].addr,
+                    uint96(_deposit.amount),
+                    new bytes(0)
+                );
+                numTransactionsConsumed++;
+                // Total balance in this contract decreases by the amount deposited
+                totalBalance[tokens[i].addr] = totalBalance[tokens[i].addr].sub(amount);
+            } else if (initialAmmBalances[i] > ammBalances[i]) {
+                uint amount = initialAmmBalances[i] - ammBalances[i];
+                // Check that the withdrawal in the block matches the expected withdrawal
+                WithdrawTransaction.Withdrawal memory withdrawal = _block.readWithdrawal(txIdx++);
+                require(withdrawal.owner == address(this), "INVALID_TX_DATA");
+                require(withdrawal.accountID == accountID, "INVALID_TX_DATA");
+                require(withdrawal.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+                require(withdrawal.amount == amount, "INVALID_TX_DATA");
+                require(withdrawal.feeTokenID == withdrawal.tokenID, "INVALID_TX_DATA");
+                require(withdrawal.fee == 0, "INVALID_TX_DATA");
+                require(withdrawal.to == address(this), "INVALID_TX_DATA");
+                require(withdrawal.extraData.length == 0, "INVALID_TX_DATA");
+                // Now approve this withdrawal
+                withdrawal.validUntil = 0xffffffff;
+                bytes32 txHash = WithdrawTransaction.hashTx(exchangeDomainSeparator, withdrawal);
+                exchange.approveTransaction(address(this), txHash);
+                numTransactionsConsumed++;
+                // Total balance in this contract increases by the amount withdrawn
+                totalBalance[tokens[i].addr] = totalBalance[tokens[i].addr].add(amount);
+            }
```

Comment:
Let me know if my understanding is correct:

1. The relayer determines how many queued Join/Exit requests to process. Once the starting and ending indices are chosen, the relayer knows the new balances in this AMM account based on the most up-to-date L2 balances.

1. But before processing the Joins/Exists, the relayer must roll up the most up-to-date L2 balance per token using an AmmUpdate transaction (per token). These **n** (*n is the number of tokens in the pool*)  transactions are put in the block next to each other.

1. The relayer then includes (up to) **n** Deposits/Withdrawals into the block in case any token's balance changed. Same as the AmmUpdate transactions, these deposit/withdraw transactions don't exist yet, but the relayer is assured that once notified, the AMM contract will create these requests onchain before block submission. If a token's balance doesn't change at all (the aggregated deposit amount == the aggregated exit amount), then there is no deposit/withdrawal for the corresponding token.

The AMM contract must be notified with the AmmUpdate transactions first, so it will be able to calculate (in the same way as the relayer) how much to deposit/withdraw per token, so it can authorize deposit/withdrawal requests onchain. It is also important for the AMM contract to min/burn ownership tokens.

Reply:
All correct. :)